File: packages/fhir-types/src/fhir_types/datatypes.py
"""FHIR R4 general-purpose datatypes.

These are the reusable elements the seven modelled resources are built from. They
are not resources: none carries a ``resourceType`` and none can be read or written
against a FHIR server on its own.

**Dates and times are ``str``, not ``date``/``datetime``.** FHIR's ``date`` and
``dateTime`` permit reduced precision — ``"2024"`` and ``"2024-03"`` are both legal
values for a birth date, and real EHR data contains them. ``datetime.date`` cannot
represent either, so parsing into it would reject valid payloads or silently invent
a day. The values stay as the server sent them; a caller that needs a real date
parses it at the point of use, where it can decide what a partial date means.
"""

from __future__ import annotations

from .base import FHIRBase
from .codes import (
    AddressType,
    AddressUse,
    ContactPointSystem,
    ContactPointUse,
    IdentifierUse,
    NameUse,
    QuantityComparator,
)


class Coding(FHIRBase):
    """A single code drawn from a code system.

    Attributes:
        system: URI of the code system, e.g. ``http://snomed.info/sct``.
        version: Version of the code system, when the code's meaning depends on it.
        code: The symbol itself, in syntax defined by the system.
        display: Human-readable label for the code, as the system defines it.
        user_selected: True when a human picked this code directly.
    """

    system: str | None = None
    version: str | None = None
    code: str | None = None
    display: str | None = None
    user_selected: bool | None = None


class CodeableConcept(FHIRBase):
    """A concept expressed as codes from one or more systems, plus free text.

    Most clinically meaningful fields are this rather than a bare ``Coding``: the
    same diagnosis routinely arrives coded in both ICD-10 and SNOMED, and payers
    differ on which they will accept.

    Attributes:
        coding: Equivalent codings of the same concept across systems.
        text: The concept as the source presented it to a human.
    """

    coding: list[Coding] | None = None
    text: str | None = None


class Period(FHIRBase):
    """A time range bounded by start and end.

    Attributes:
        start: Inclusive start, as a ``dateTime`` string.
        end: Inclusive end. Absent means the period has no known end.
    """

    start: str | None = None
    end: str | None = None


class Identifier(FHIRBase):
    """A business identifier — an MRN, a policy number, a claim number.

    Attributes:
        use: Role this identifier plays for its owner.
        type: Coded description of the identifier's kind.
        system: Namespace the value is unique within.
        value: The identifier itself. Frequently PHI (an MRN or member id).
        period: When the identifier was or is valid.
        assigner: Organization that issued it.
    """

    use: IdentifierUse | None = None
    type: CodeableConcept | None = None
    system: str | None = None
    value: str | None = None
    period: Period | None = None
    assigner: Reference | None = None


class Reference(FHIRBase):
    """A pointer from one resource to another.

    Attributes:
        reference: Relative or absolute URL, e.g. ``Patient/1234``.
        type: Resource type being referred to, when ``reference`` is absent.
        identifier: Business identifier for the target, when it has no URL.
        display: Human-readable label for the target.
    """

    reference: str | None = None
    type: str | None = None
    identifier: Identifier | None = None
    display: str | None = None


class HumanName(FHIRBase):
    """A name of a human, with the parts kept separate. Always PHI.

    Attributes:
        use: Role this name plays — official, nickname, maiden, and so on.
        text: The full name as it should be displayed.
        family: Family name (surname).
        given: Given and middle names, in order.
        prefix: Titles preceding the name, e.g. ``Dr``.
        suffix: Qualifiers following the name, e.g. ``Jr``, ``MD``.
        period: When this name was or is in use.
    """

    use: NameUse | None = None
    text: str | None = None
    family: str | None = None
    given: list[str] | None = None
    prefix: list[str] | None = None
    suffix: list[str] | None = None
    period: Period | None = None


class ContactPoint(FHIRBase):
    """A phone number, email address or other contact detail. Always PHI.

    Attributes:
        system: Which kind of contact channel this is.
        value: The number or address itself.
        use: The context this channel is used in.
        rank: Preference order, 1 being most preferred.
        period: When this channel was or is in use.
    """

    system: ContactPointSystem | None = None
    value: str | None = None
    use: ContactPointUse | None = None
    rank: int | None = None
    period: Period | None = None


class Address(FHIRBase):
    """A postal address. Always PHI.

    Attributes:
        use: The purpose this address serves.
        type: Whether the address is postal, physical, or both.
        text: The address as a single displayable string.
        line: Street lines, in order.
        city: City or town.
        district: County or district.
        state: State or province — payer policies are state-scoped, so this drives
            the ``state`` component of the RAG cache key described in CLAUDE.md.
        postal_code: Postal or ZIP code.
        country: Country, as a name or ISO 3166 code.
        period: When this address was or is in use.
    """

    use: AddressUse | None = None
    type: AddressType | None = None
    text: str | None = None
    line: list[str] | None = None
    city: str | None = None
    district: str | None = None
    state: str | None = None
    postal_code: str | None = None
    country: str | None = None
    period: Period | None = None


class Attachment(FHIRBase):
    """Content referred to in-line or by URL.

    Attributes:
        content_type: MIME type of the content, including character encoding.
        language: Language of the content, as a BCP-47 tag.
        data: Base64 of the content itself. On a DocumentReference this is the note
            body — PHI, and never safe to log or put in an error message.
        url: Location the content can be retrieved from instead.
        size: Size in bytes of the decoded content.
        title: Label for the attachment.
        creation: When the attachment was first created, as a ``dateTime``.
    """

    content_type: str | None = None
    language: str | None = None
    data: str | None = None
    url: str | None = None
    size: int | None = None
    title: str | None = None
    creation: str | None = None


class Quantity(FHIRBase):
    """A measured or counted amount with a unit.

    Attributes:
        value: The numeric value.
        comparator: Set when the value is a bound rather than a measurement.
        unit: Human-readable unit.
        system: URI of the unit system, usually UCUM.
        code: Machine-readable form of the unit.
    """

    value: float | None = None
    comparator: QuantityComparator | None = None
    unit: str | None = None
    system: str | None = None
    code: str | None = None


class Money(FHIRBase):
    """An amount of currency.

    Attributes:
        value: The numeric amount.
        currency: ISO 4217 currency code.
    """

    value: float | None = None
    currency: str | None = None


class Annotation(FHIRBase):
    """A free-text note attached to a resource, with its author and time.

    ``text`` is clinician-authored prose and is PHI in every case this project
    encounters it.

    Attributes:
        author_reference: The author, as a reference to a Practitioner or Patient.
        author_string: The author's name, when there is no resource for them.
        time: When the note was made, as a ``dateTime``.
        text: The note itself.
    """

    author_reference: Reference | None = None
    author_string: str | None = None
    time: str | None = None
    text: str


# Identifier.assigner forward-references Reference, which in turn holds an
# Identifier. The cycle is legal but has to be resolved once both exist.
Identifier.model_rebuild()

File: packages/fhir-types/typescript/src/datatypes.ts
/**
 * FHIR R4 general-purpose datatypes, mirroring `src/fhir_types/datatypes.py`.
 *
 * Dates and times are `string`, not `Date`. FHIR's `date` and `dateTime` permit
 * reduced precision — `"2024"` and `"2024-03"` are both legal birth dates and real
 * EHR data contains them. `new Date("2024-03")` silently invents a day, so the
 * values stay as the server sent them and a caller parses at the point of use.
 */

import type {
  AddressType,
  AddressUse,
  ContactPointSystem,
  ContactPointUse,
  IdentifierUse,
  NameUse,
  QuantityComparator,
} from './codes.js';

/** A single code drawn from a code system. */
export interface Coding {
  /** URI of the code system, e.g. `http://snomed.info/sct`. */
  readonly system?: string;
  /** Version of the code system, when the code's meaning depends on it. */
  readonly version?: string;
  /** The symbol itself, in syntax defined by the system. */
  readonly code?: string;
  /** Human-readable label for the code, as the system defines it. */
  readonly display?: string;
  /** True when a human picked this code directly. */
  readonly userSelected?: boolean;
}

/**
 * A concept expressed as codes from one or more systems, plus free text.
 *
 * Most clinically meaningful fields are this rather than a bare `Coding`: the same
 * diagnosis routinely arrives coded in both ICD-10 and SNOMED, and payers differ
 * on which they accept.
 */
export interface CodeableConcept {
  /** Equivalent codings of the same concept across systems. */
  readonly coding?: readonly Coding[];
  /** The concept as the source presented it to a human. */
  readonly text?: string;
}

/** A time range bounded by start and end. */
export interface Period {
  /** Inclusive start, as a `dateTime` string. */
  readonly start?: string;
  /** Inclusive end. Absent means the period has no known end. */
  readonly end?: string;
}

/** A business identifier — an MRN, a policy number, a claim number. */
export interface Identifier {
  /** Role this identifier plays for its owner. */
  readonly use?: IdentifierUse;
  /** Coded description of the identifier's kind. */
  readonly type?: CodeableConcept;
  /** Namespace the value is unique within. */
  readonly system?: string;
  /** The identifier itself. Frequently PHI (an MRN or member id). */
  readonly value?: string;
  /** When the identifier was or is valid. */
  readonly period?: Period;
  /** Organization that issued it. */
  readonly assigner?: Reference;
}

/** A pointer from one resource to another. */
export interface Reference {
  /** Relative or absolute URL, e.g. `Patient/1234`. */
  readonly reference?: string;
  /** Resource type being referred to, when `reference` is absent. */
  readonly type?: string;
  /** Business identifier for the target, when it has no URL. */
  readonly identifier?: Identifier;
  /** Human-readable label for the target. */
  readonly display?: string;
}

/** A name of a human, with the parts kept separate. Always PHI. */
export interface HumanName {
  /** Role this name plays — official, nickname, maiden, and so on. */
  readonly use?: NameUse;
  /** The full name as it should be displayed. */
  readonly text?: string;
  /** Family name (surname). */
  readonly family?: string;
  /** Given and middle names, in order. */
  readonly given?: readonly string[];
  /** Titles preceding the name, e.g. `Dr`. */
  readonly prefix?: readonly string[];
  /** Qualifiers following the name, e.g. `Jr`, `MD`. */
  readonly suffix?: readonly string[];
  /** When this name was or is in use. */
  readonly period?: Period;
}

/** A phone number, email address or other contact detail. Always PHI. */
export interface ContactPoint {
  /** Which kind of contact channel this is. */
  readonly system?: ContactPointSystem;
  /** The number or address itself. */
  readonly value?: string;
  /** The context this channel is used in. */
  readonly use?: ContactPointUse;
  /** Preference order, 1 being most preferred. */
  readonly rank?: number;
  /** When this channel was or is in use. */
  readonly period?: Period;
}

/** A postal address. Always PHI. */
export interface Address {
  /** The purpose this address serves. */
  readonly use?: AddressUse;
  /** Whether the address is postal, physical, or both. */
  readonly type?: AddressType;
  /** The address as a single displayable string. */
  readonly text?: string;
  /** Street lines, in order. */
  readonly line?: readonly string[];
  /** City or town. */
  readonly city?: string;
  /** County or district. */
  readonly district?: string;
  /** State or province. **Not** the source of the `state` component of the RAG
   * cache key, whatever resource this address hangs off. An earlier version of
   * this comment said it was, written before anyone read what the policy
   * documents say about their own applicability. They say the site of care —
   * CMS's Medicare Coverage Database instructs the reader to select "the state
   * where the service took place" — so the key's `state` comes from the
   * encounter's Location/Organization address, never from a `Patient.address`.
   * See TASK-052b, "Where `state` comes from". */
  readonly state?: string;
  /** Postal or ZIP code. */
  readonly postalCode?: string;
  /** Country, as a name or ISO 3166 code. */
  readonly country?: string;
  /** When this address was or is in use. */
  readonly period?: Period;
}

/** Content referred to in-line or by URL. */
export interface Attachment {
  /** MIME type of the content, including character encoding. */
  readonly contentType?: string;
  /** Language of the content, as a BCP-47 tag. */
  readonly language?: string;
  /** Base64 of the content itself. On a DocumentReference this is the note body —
   * PHI, and never safe to log or put in an error message. */
  readonly data?: string;
  /** Location the content can be retrieved from instead. */
  readonly url?: string;
  /** Size in bytes of the decoded content. */
  readonly size?: number;
  /** Label for the attachment. */
  readonly title?: string;
  /** When the attachment was first created, as a `dateTime`. */
  readonly creation?: string;
}

/** A measured or counted amount with a unit. */
export interface Quantity {
  /** The numeric value. */
  readonly value?: number;
  /** Set when the value is a bound rather than a measurement. */
  readonly comparator?: QuantityComparator;
  /** Human-readable unit. */
  readonly unit?: string;
  /** URI of the unit system, usually UCUM. */
  readonly system?: string;
  /** Machine-readable form of the unit. */
  readonly code?: string;
}

/** An amount of currency. */
export interface Money {
  /** The numeric amount. */
  readonly value?: number;
  /** ISO 4217 currency code. */
  readonly currency?: string;
}

/**
 * A free-text note attached to a resource, with its author and time.
 *
 * `text` is clinician-authored prose and is PHI in every case this project
 * encounters it.
 */
export interface Annotation {
  /** The author, as a reference to a Practitioner or Patient. */
  readonly authorReference?: Reference;
  /** The author's name, when there is no resource for them. */
  readonly authorString?: string;
  /** When the note was made, as a `dateTime`. */
  readonly time?: string;
  /** The note itself. */
  readonly text: string;
}
